### ISSUE TITLE
contains() no longer works with lists of file objects

### ISSUE THREAD

**Issue by MathieuDuponchelle** [mesonbuild/meson#9020]
**Describe the bug**

With 0.59.0, `[].contains()` no longer works as expected with file objects

**To Reproduce**

foo.c:

```
```
meson.build:

```
project('contains-bug', 'c',
  meson_version: '>= 0.59.0',
)

l = [files('foo.c')]

message(l.contains(files('foo.c')))
```

**Expected behavior**

This should print `Message: True`

**system parameters**

* plain native build
* Fedora 33
* python 3.9.2
* meson 0.59.0
* ninja 1.10.2

**Comment by xclaesse** [comment]
`l` is empty, so contains() always should return false. Maybe a typo in your issue report, if I change to `f.contains()` then it prints False but I guess it should be True.

**Comment by MathieuDuponchelle** [comment]
hah, my bad, edited

**Comment by MathieuDuponchelle** [comment]
for context, this is in gstreamer's FFmpeg build, where we relied on the feature for poor man's deduplication: https://gitlab.freedesktop.org/gstreamer/meson-ports/ffmpeg/-/blob/meson-4.3/meson.build#L2118

**Comment by xclaesse** [comment]
Ok, I can confirm that with list of strings it works, but not with File objects.

The test case in this report could be a bit confusing because `files()` returns itself a list, so there is also the question if `[files()]` should be flattened. Also raise the question if calling `files()` twice for the same string should return the same object.

A simplified test case:
```
l = files('foo.c')
assert(l.contains(l[0]))
```

@mensinda that smell like a regression in holderify stuff you did, is it possible?

**Comment by xclaesse** [comment]
@MathieuDuponchelle I guess this is about the duplicate `g722dsp.o` ninja rule issue in FFmpeg we discussed with @tp-m on IRC yesterday?

**Comment by MathieuDuponchelle** [comment]
@xclaesse I didn't see the discussion, but got a private report today about it yes :)

**Comment by mensinda** [comment]
@xclaesse Maybe. The relevant code is [here](https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreterbase/interpreterbase.py#L786-L798) and from the looks of it flattening `posargs[0]` should fix the issue (and possibly introduce others...). Though, I honestly have no idea how this used to work previeously...

It is also slightly concerning that we use `==` for the check and I am not sure if we implement the `__eq__` function for all interpreter objects...

**Comment by MathieuDuponchelle** [comment]
Putting flattening aside, this also prints False when I would expect it to print True (and that more closely matches the pattern in our FFmpeg meson.build):

```
project('contains-bug', 'c',
  meson_version: '>= 0.59.0',
)

fs = import('fs')

l = []

foreach f: files('foo.c')
  l += [f]
endforeach

foreach f: files('foo.c')
  message(l.contains(f))
endforeach
```

**Comment by mensinda** [comment]
Should be fixed with #9027

**Comment by MathieuDuponchelle** [comment]
well done, thanks folks :)
